### PR TITLE
ToolsetReader: Fixes and code improvements for mono build

### PR DESCRIPTION
**Pull request by mfilippov**
I fix build on Mono

### CODE REVIEW

**Comment by mfilippov** [comment]
I see problem in unit tests I try fix it.

**Comment by rainersigwald** [comment]
Thanks! The I'm filing bugs for all of the not-related-to-your-change test failures.

Can you amend the GitVersioning commit so that it mentions it's a workaround for #694 (that I just filed)? That way we can be sure to unify the logic later.

**Comment by rainersigwald** [comment]
(Sorry, you may have beat me to it!)

**Comment by rainersigwald** [comment]
#695 is the Windows failure

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 1]
```diff
@@ -204,7 +204,8 @@ ToolsetDefinitionLocations locations
 
                         // Other toolsets are installed in the xbuild directory
                         var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
-                        var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
+                        var regexPath = xbuildToolsetsDir + Path.DirectorySeparatorChar;
+                        var r = new Regex(regexPath.Replace(@"\", @"\\") + @"\d+\.\d+");
```
Comment (rainersigwald) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Can you use `Regex.Escape` instead?

**Comment by rainersigwald** [comment]
Fantastic, thanks! Looks good to me, once you've changed the commit message for the GitVersioning commit.

**Comment by mfilippov** [comment]
Commit message changed.

**Comment by mfilippov** [comment]
I little change GitVersioning workaround for include targets only if it exists.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 2]
```diff
@@ -153,23 +153,26 @@ ToolsetDefinitionLocations locations
...
-
-                // We do not accumulate properties when reading them from the registry, because the order
-                // in which values are returned to us is essentially random: so we disallow one property
-                // in the registry to refer to another also in the registry
-                defaultToolsVersionFromRegistry = registryReader.ReadToolsets(toolsets, globalProperties,
-                    initialProperties, false /* do not accumulate properties */, out overrideTasksPathFromRegistry,
-                    out defaultOverrideToolsVersionFromRegistry);
-            }
+
+                if (NativeMethodsShared.IsWindows || registryReader != null)
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Is this runtime `IsWindows` check required here, since `FEATURE_WIN32_REGISTRY` should be enabled only for windows/full framework builds anyway? And line 163 (`registryReader = registryReader ?? new ToolsetRegistryReader(...` means that we don't need to check for `registryReader != null` either.

And so, the new set of braces being added here are not required.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 3]
```diff
@@ -240,8 +243,9 @@ ToolsetDefinitionLocations locations
                         }
                     }
                 }
+#if FEATURE_WIN32_REGISTRY
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Because of the comment above, this should not be required either, just drop the extra brace.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 4]
```diff
@@ -205,42 +205,45 @@ ToolsetDefinitionLocations locations
                         // Other toolsets are installed in the xbuild directory
                         var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Nitpick - maybe we can add the `+ Path.DirectorySeparatorChar` here, instead of later. So, it's use with `Regex.Escape` later becomes easier to read.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 5]
```diff
@@ -205,42 +205,45 @@ ToolsetDefinitionLocations locations
...
+                                            buildProperties,
+                                            environmentProperties,
+                                            globalProperties,
+                                            null,
+                                            currentDir,
+                                            string.Empty));
+                                }
+                            }
 
+                            if (!toolsets.ContainsKey(MSBuildConstants.CurrentToolsVersion))
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Does this `if` block need to be under the directory check? IIUC, this should be run irrespective of the earlier block.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 6]
```diff
@@ -204,7 +204,7 @@ ToolsetDefinitionLocations locations
 
                         // Other toolsets are installed in the xbuild directory
                         var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
-                        var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
+                        var r = new Regex(Regex.Escape(xbuildToolsetsDir + Path.DirectorySeparatorChar) + @"\d+\.\d+");
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Could you please squash this with the related commit in the `ToolsetReader`?

Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Though, maybe it is good as a separate commit. Just add some reasoning for the change, like add the stack trace from the issue in the commit message here.

Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: And you could move that `var r = new Regex...` inside the directory.exists block.

**Comment by radical** [comment]
Also, could you please add more information/explanation to the commit messages like for `Check folder exists` and `fix path escaping...` .

And thanks for your PR :)

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs (hunk 2 above)
Comment (mfilippov) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: If NativeMethodsShared.IsWindows but registryReader == null conditions is true and registryReader = registryReader ?? new ToolsetRegistryReader.. required. I little rewrite method for remove else. I commit updated version soon.

**Comment by mfilippov** [comment]
All code review comments fixed.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs (hunk 2 above)
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Ah, my comment was incorrect. `FEATURE_WIN32_REGISTRY` will be enabled even for mono builds, which can run on non-windows platforms too. So, the runtime check `IsWindows` is useful here.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 7]
```diff
@@ -157,89 +157,86 @@ ToolsetDefinitionLocations locations
             {
 #if FEATURE_WIN32_REGISTRY
                 if (NativeMethodsShared.IsWindows || registryReader != null)
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Since, this check is indeed required (https://github.com/Microsoft/msbuild/pull/693#discussion_r68432809), the only thing required to be fixed here is the opening brace for this `if` block. And just indent the corresponding code. Something like - https://gist.github.com/radical/f085ffb25f57b6fc179f785e4a47b930 .

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 8]
```diff
@@ -117,7 +117,7 @@ internal static string ReadAllToolsets
             ToolsetDefinitionLocations locations
             )
         {
-            PropertyDictionary<ProjectPropertyInstance> initialProperties =
+            var initialProperties =
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Please don't add unrelated changes to the commit.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 9]
```diff
@@ -157,89 +157,86 @@ ToolsetDefinitionLocations locations
...
-                                    string.Empty));
-                        }
+                    // Other toolsets are installed in the xbuild directory
+                    var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
+                    var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
+                    foreach (var d in Directory.GetDirectories(xbuildToolsetsDir).Where(d => r.IsMatch(d)))
+                    {
+                        var version = Path.GetFileName(d);
+                        var binPath = Path.Combine(d, "bin");
+                        if (toolsets.ContainsKey(version))
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: This change here was easy to miss, because of the big diff and most of it being just "indentation change". This commit is fixing the "merge" issue, so, it should not contain these unrelated changes. It can be fixed in the separate commit where you are fixing the Regex issue.

File: src/XMakeBuildEngine/Definition/ToolsetReader.cs [hunk 10]
```diff
@@ -199,29 +199,32 @@ ToolsetDefinitionLocations locations
                     }
 
                     // Other toolsets are installed in the xbuild directory
-                    var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
-                    var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
-                    foreach (var d in Directory.GetDirectories(xbuildToolsetsDir).Where(d => r.IsMatch(d)))
+                    var xbuildToolsetsDir = Path.Combine(libraryPath, $"xbuild{Path.DirectorySeparatorChar}");
```
Comment (radical) [inline src/XMakeBuildEngine/Definition/ToolsetReader.cs]: Commit message should explain the reasoning for the changes. What is being changed can be seen in the commit itself, but the reason is important and can be useful in debugging or understanding it in future. So, something like this:

```
ToolsetReader: Escape paths before using them as part of a regex

We are trying to find toolset paths for mono, and using regex to match such
paths to a pattern. But this fails when running on windows and the path contains
`\` like `D:\j\workspace\_Windows_NT`, because `Regex` interprets it as a escape
sequence. It manifests as:

MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
11:59:04 System.ArgumentException: parsing 'D:\j\workspace\_Windows_NT_p---ab87db35\bin\xbuild\\d+\.\d+' - Unrecognized escape sequence \\j.
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanCharEscape()
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanBasicBackslash()
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanRegex()
11:59:04    at System.Text.RegularExpressions.RegexParser.Parse(String re, RegexOptions op)
11:59:04    at System.Text.RegularExpressions.Regex..ctor(String pattern, RegexOptions options, TimeSpan matchTimeout, Boolean useCache)
11:59:04    at System.Text.RegularExpressions.Regex..ctor(String pattern)
11:59:04    at Microsoft.Build.Evaluation.ToolsetReader.ReadAllToolsets(Dictionary`2 toolsets, PropertyDictionary`1 environmentProperties, PropertyDictionary`1 globalProperties, ToolsetDefinitionLocations locations)
11:59:04    at Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection()
11:59:04    at Microsoft.Build.Evaluation.ProjectCollection..ctor(IDictionary`2 globalProperties, IEnumerable`1 loggers, IEnumerable`1 remoteLoggers, ToolsetDefinitionLocations toolsetDefinitionLocations, Int32 maxNodeCount, Boolean onlyLogCriticalEvents)
11:59:04    at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, Boolean debugger, Boolean detailedSummary)
11:59:04    at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)

Use `Regex.Escape` for the path portion of the pattern.

Also, see https://github.com/Microsoft/msbuild/issues/695
```

**Comment by AndyGerlicher** [comment]
LGTM, but can you drop your commit about the Git Versioning? #721 should solve that issue for now.

**Comment by mfilippov** [comment]
@AndyGerlicher Git Versioning commit dropped, @radical I try rearrange PR base on you comments.

**Comment by radical** [comment]
If the ToolsetReader commits are being retained in this PR itself, then do update the PR commit message to indicate that this is more than just build fix. Other than that, LGTM!

**Comment by rainersigwald** [comment]
Great, thank you @mfilippov!
